azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message.py

```diff
@@ -0,0 +1,53 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class Message(object):
+    def __init__(
+        self, data, creation_time_utc, message_id=None, content_encoding=None, content_type=None
+    ):
+        """
+        :param data: The  data that constitutes the payload
+        :param creation_time_utc: Date and time the message was created
+        :param message_id: A user-settable identifier for the message used for request-reply patterns.Format: A case-sensitive string (up to 128 characters long) of ASCII 7-bit alphanumeric characters + {'-', ':', '.', '+', '%', '_', '#', '*', '?', '!', '(', ')', ',', '=', '@', ';', '$', '''}.
+        :param content_encoding: Content encoding of the message data. can be 'utf-8', 'utf-16' or 'utf-32'.
+        :param content_type: Content type property used to routes with the message body. Should be 'application/json'.
+        """
+        self.data = data
+        self.creation_time_utc = creation_time_utc
+        self.custom_properties = {}
+        self.lock_token = None
+        self.message_id = message_id
+        self.sequence_number = None
+        self.to = None
+        self.expiry_time_utc = None
+        self.enqueued_time = None
+        self.correlation_id = None
+        self.user_id = None
+        self.ack = None
+        self.content_encoding = content_encoding
+        self.content_type = content_type
+        self.device_id = None
+        self.auth_generation_id = None
+        self.auth_method = None
+
+    def add_property(self, key, value):
```

Comment:
why not simply let the user set custom_properties by accessing the dictionary directly? after all, it's public. this helper method just gives them another way to do it. Then in the transport, where there might be some limitation on property names, we throw about them when we translate them. 

---
In reply to: [237663623](https://github.com/Azure/azure-iot-sdk-python-preview/pull/27#discussion_r237663623) [](ancestors = 237663623)

Reply:
I am gone between the 2 names. I am fine if we want to call it `user_properties` .. C# calls it that but also C# also has another dictionary called `system_properties`. Node calls it just properties. Have changed the name to `add_custom_property`  Updated the documentation which explicitly can call out that we can add to some user defined properties. 

---
In reply to: [237663623](https://github.com/Azure/azure-iot-sdk-python-preview/pull/27#discussion_r237663623) [](ancestors = 237663623)

Reply:
what's the value of adding a method instead of accessing the (public) dictionary object?

---
In reply to: [237723191](https://github.com/Azure/azure-iot-sdk-python-preview/pull/27#discussion_r237723191) [](ancestors = 237723191,237663623)